src/meta/src/hummock/manager/compaction_group_manager.rs

```diff
@@ -695,113 +714,100 @@ impl HummockManager {
     default_config: CompactionConfig,
     /// Tables that write limit is trigger for.
     pub write_limit: HashMap<CompactionGroupId, WriteLimit>,
-    meta_store_impl: MetaStoreImpl,
 }
 
 // init method
 impl CompactionGroupManager {
-    async fn init(&mut self) -> Result<()> {
-        let loaded_compaction_groups: BTreeMap<CompactionGroupId, CompactionGroup> =
-            match &self.meta_store_impl {
-                MetaStoreImpl::Kv(meta_store) => CompactionGroup::list(meta_store)
-                    .await?
-                    .into_iter()
-                    .map(|cg| (cg.group_id(), cg))
-                    .collect(),
-                MetaStoreImpl::Sql(sql_meta_store) => {
-                    use sea_orm::EntityTrait;
-                    compaction_config::Entity::find()
-                        .all(&sql_meta_store.conn)
-                        .await
-                        .map_err(MetadataModelError::from)?
-                        .into_iter()
-                        .map(|m| (m.compaction_group_id as CompactionGroupId, m.into()))
-                        .collect()
-                }
-            };
+    fn init(&mut self, loaded_compaction_groups: BTreeMap<CompactionGroupId, CompactionGroup>) {
         if !loaded_compaction_groups.is_empty() {
             self.compaction_groups = loaded_compaction_groups;
         }
-        Ok(())
-    }
-
-    /// Initializes the config for a group.
-    /// Should only be used by compaction test.
-    pub(super) async fn init_compaction_config_for_replay(
-        &mut self,
-        group_id: CompactionGroupId,
-        config: CompactionConfig,
-    ) -> Result<()> {
-        let insert = BTreeMapEntryTransaction::new_insert(
-            &mut self.compaction_groups,
-            group_id,
-            CompactionGroup {
-                group_id,
-                compaction_config: Arc::new(config),
-            },
-        );
-        commit_multi_var!(self.meta_store_impl, insert)?;
-        Ok(())
     }
 }
 
 impl CompactionGroupManager {
-    /// Gets compaction group config for `compaction_group_id`, inserts default one if missing.
-    async fn get_or_insert_compaction_group_config(
-        &mut self,
-        compaction_group_id: CompactionGroupId,
-    ) -> Result<CompactionGroup> {
-        let r = self
-            .get_or_insert_compaction_group_configs(&[compaction_group_id])
-            .await?;
-        Ok(r.into_values().next().unwrap())
+    pub fn start_compaction_groups_txn(&mut self) -> BTreeMapTransaction<'_, u64, CompactionGroup> {
+        BTreeMapTransaction::new(&mut self.compaction_groups)
     }
 
-    /// Gets compaction group configs for `compaction_group_ids`, inserts default one if missing.
-    async fn get_or_insert_compaction_group_configs(
-        &mut self,
+    /// Inserts compaction group configs if they do not exist.
+    pub fn insert_not_exist_groups_with_config(
```

Comment:
Thanks for the effort of refactoring the codes. There are a few things I think that can be improved for the interface design (the interfaces are kind of messy before so not all are related to changes under this PR):
1. `BTreeMapTransaction` but not `self` is involved in almost all interfaces in `CompactionGroupManager`, which indicates that these interfaces should be put as part of `CompactionGroupManager`
2. `try_get_compaction_group_config` takes a `&BTreeMap` and `&BTreeMap` comes from the `pub compaction_groups` in `CompactionGroupManager`, which seems very strange. We should keep `compaction_groups` private and let `try_get_compaction_group_config` takes `&self` instead.
3. `default_config` is being used both by directly accessing the struct member and via `fn default_compaction_config`. We should unify the usage by only accessing the default config via `fn default_compaction_config`.
4. `default_config` will never be changes. I suggest we wrap it in `Arc` so that we don't need to deep copy it every time in `fn default_compaction_config`.


```rust
impl<'a> BTreeMapTransaction<'a, u64, CompactionGroup> {
    pub fn try_get_compaction_group_config(&self, compaction_group_id: u64) -> Option<CompactionGroup> {
        ...
    }

    // Read uncommitted
    pub fn try_create_compaction_group(&mut self, compaction_group_ids: &[CompactionGroupId], config: CompactionConfig) -> bool {
        ...
    }

    pub fn update_compaction_groups_config(
        &mut self,
        compaction_group_ids: &[CompactionGroupId],
        config_to_update: &[MutableConfig],
    ) {
        ...
    }

    pub fn purge(&mut self,existing_groups: HashSet<CompactionGroupId>) {
        ...
    }
}

struct CompactionGroupManager {
    compaction_groups: BTreeMap<CompactionGroupId, CompactionGroup>,
    default_config: Arc<CompactionConfig>,
    /// Tables that write limit is trigger for.
    pub write_limit: HashMap<CompactionGroupId, WriteLimit>,
}

impl CompactionGroupManager {
    pub fn start_compaction_groups_txn(&mut self) -> BTreeMapTransaction<'_, u64, CompactionGroup> {
        ...
    }

    // Read committed
    pub(super) fn try_get_compaction_group_config(&self, compaction_group_id: CompactionGroupId) -> Option<CompactionGroup>
    {
        ...
    }

    pub(super) fn default_compaction_config(&self) -> &Arc<CompactionConfig> {
        self.default_config.clone()
    }
}
```